Add DRA snapshot build

breskeby opened the pull request:
related add support to run build against a fixed buildIds for ml-cpp and beats

breskeby committed:
Add support for building against defined build-ids of dependent artifacts

breskeby committed:
Simplify resolve dra buildid script

breskeby committed:
Cleanup buildIds resolution script

breskeby committed:
Introduce DraResolvePlugin and tests

breskeby committed:
Rework DRA testing

breskeby committed:
Support beats like multi artifact dra artifacts

breskeby committed:
Apply DRA binary plugin in distro and ml project

breskeby committed:
Allow resolve buildIds from sysprops

breskeby committed:
Declare initial dra snapshot ci job

breskeby committed:
First version of DRA snapshot ci job

mark-vieira commented on .ci/jobs.t/elastic+elasticsearch+periodic+dra-snapshots-trigger.yml:
```diff
@@ -0,0 +1,6 @@
+---
+jjbb-template: periodic-trigger-lgc.yml
```
We can start with this, but I think we'll want to publish snapshots on every successful intake commit. I don't see any reason not to.

breskeby replied:
I'm happy with every successful commit. I just used that as it is DAYLY releasable artifacts. but maybe took it to literately :)

mark-vieira replied:
_At least_ daily 😉

mark-vieira commented on .ci/jobs.t/elastic+elasticsearch+periodic+dra-snapshots.yml:
```diff
@@ -0,0 +1,45 @@
+            RUNTIME_JAVA_HOME=$HOME/.java/$ES_RUNTIME_JAVA
+      - shell: |
+            #!/usr/local/bin/runbld --redirect-stderr
+            DRA_VERSION=`cat build-tools-internal/version.properties \
```
I think we should standardize on a syntax here. Some quick googling revealed that the backtick syntax is deprecated and `$()` is prefered for command evaluation. I suggest we use that across the board in our shell scripts.

breskeby replied:
done

mark-vieira commented on .ci/jobs.t/elastic+elasticsearch+periodic+dra-snapshots.yml:
```diff
@@ -0,0 +1,45 @@
+                  | sed "s/elasticsearch//g" \
+                  | sed "s/ //g" \
+                  | sed "s/=//"`
+            beatsBuildId=`./.ci/scripts/resolve-dra-manifest.sh beats master`
```
Should we use the convention of upper-snake-case for these instead of camel-case?

breskeby replied:
fixed

mark-vieira commented on .ci/jobs.t/elastic+elasticsearch+periodic+dra-snapshots.yml:
```diff
@@ -0,0 +1,45 @@
+                   cli collect \
+                   --project elasticsearch \
+                   --branch "master" \
+                   --commit `git rev-parse HEAD` \
```
We can just use the `GIT_COMMIT` env var here provided by Jenkins.

breskeby replied:
fixed

mark-vieira commented on .ci/jobs.t/elastic+elasticsearch+periodic+dra-snapshots.yml:
```diff
@@ -0,0 +1,45 @@
+                   --workflow "snapshot" \
+                   --version "$DRA_VERSION" \
+                   --artifact-set main \
+                   --dependency https://artifacts-snapshot.elastic.co/beats/$(echo $beatsBuildId | sed "s/beats=//g")/manifest-${DRA_VERSION}-SNAPSHOT.json \
```
Maybe we should do the `sed` bit up above, or in the resolve-dra-manifest script itself.

breskeby replied:
I'll move the url calculation up

breskeby replied:
reworked this

mark-vieira commented on .ci/jobs.t/elastic+elasticsearch+periodic+dra-snapshots.yml:
```diff
@@ -0,0 +1,45 @@
+            RUNTIME_JAVA_HOME=$HOME/.java/$ES_RUNTIME_JAVA
+      - shell: |
+            #!/usr/local/bin/runbld --redirect-stderr
+            DRA_VERSION=`cat build-tools-internal/version.properties \
```
I'd just call this `ES_VERSION` since that's what it is and there's noting "dra-specific" about how we version artifacts.

mark-vieira commented on .ci/scripts/resolve-dra-manifest.sh:
```diff
@@ -0,0 +1,14 @@
+
+curl -sS https://artifacts-snapshot.elastic.co/$ARTIFACT/latest/$BRANCH.json \
+  | grep build_id \
+  | sed "s/build_id/$ARTIFACT/g" \
```
[Jq](https://stedolan.github.io/jq/) is available on our CI agents and much better suited to parsing JSON than `sed`. I suggest we use that.

breskeby replied:
ah I wasn't sure about that and kept it simple. +1

breskeby replied:
done

mark-vieira commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/dra/DraResolvePlugin.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    private Provider<Map<String, String>> resolveBuildIdProperties() {
+        return providerFactory.systemPropertiesPrefixedBy(DRA_ARTIFACTS_DEPENDENCY_PREFIX)
```
TIL `systemPropertiesPrefixedBy` was a thing. Nice.

mark-vieira commented on distribution/docker/build.gradle:
```diff
@@ -35,34 +39,42 @@ repositories {
 testFixtures.useFixture()
 
+repositories.all {
+  println it.name
```
Let's make sure this doesn't accidentally sneak in.

breskeby replied:
removed

mark-vieira commented on build-tools/src/testFixtures/groovy/org/elasticsearch/gradle/fixtures/WiremockFixture.groovy:
```diff
@@ -67,6 +67,7 @@ class WiremockFixture {
             return buildRunClosure.call(wireMock);
         } catch (Exception e) {
             // for debugging
+            e.printStackTrace()
```
Why don't we use a proper logger here?

mark-vieira commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/dra/DraResolvePlugin.java:
```diff
@@ -0,0 +1,117 @@
+            )
+            .zip(providerFactory.provider(() -> {
+                Properties draProperties = new Properties();
+                File buildIdsFile = new File(buildLayout.getRootDirectory(), "buildIds.properties");
```
In what scenario would we get these from a properties file vs system properties? Is it necessary to support this use case?

breskeby replied:
I initially started with that and then moved to passing in via sys props. I'll delete that bit here again to keep it simple. initially thought it might be handy but not worth to keep it I guess

breskeby replied:
removed

breskeby committed:
apply review feedback

breskeby committed:
Simplify DraResolvePlugin

breskeby committed:
Fix ci script missing quotes

breskeby committed:
Cleanup job description format

breskeby committed:
Build DRAs always from ubuntu

- ubuntu was used by release manager before
- only ci that currently has libsxml2utils installed

mark-vieira commented on .ci/scripts/resolve-dra-manifest.sh:
```diff
@@ -0,0 +1,10 @@
+
+curl -sS https://artifacts-snapshot.elastic.co/$ARTIFACT/latest/$BRANCH.json \
+  | jq -r '.build_id' \
+  | xargs
```
What's the point of `xargs` here?

breskeby replied:
it was required in earlier versions of that script. removed

mark-vieira commented on .ci/scripts/resolve-dra-manifest.sh:
```diff
@@ -0,0 +1,10 @@
+  | jq -r '.build_id' \
+  | xargs
+
+# https://artifacts-snapshot.elastic.co/beats/0.1.0-283a66b2/manifest-0.1.0-SNAPSHOT.json
```
Does this comment need to remain?

breskeby replied:
removed

mark-vieira commented on build-tools-internal/src/integTest/groovy/org/elasticsearch/gradle/fixtures/LocalRepositoryFixture.groovy:
```diff
@@ -39,16 +39,21 @@ class LocalRepositoryFixture extends ExternalResource {
-                dynamicType.inject(jarFile);
-            }else {
-                dynamicType.toJar(jarFile);
+        if(clazzNames.size() == 0 ) {
```
nit: whitespace here is a bit odd

breskeby replied:
fixed

mark-vieira commented on x-pack/plugin/ml/build.gradle:
```diff
@@ -12,40 +14,44 @@ esplugin {
 }
 
-def localRepo = providers.systemProperty('build.ml_cpp.repo').orNull
+boolean useDraWithBuildId = providers.systemProperty(DraResolvePlugin.USE_DRA_ARTIFACTS_FLAG).getOrNull()
```
Should we just add this as an extension property in the `DraResolvePlugin` so we aren't having to duplicate this?

breskeby replied:
done

mark-vieira reviewed (approved):
Couple comments but LGTM. Additionaly, can we remove the ci-release.sh script?

breskeby committed:
Apply more review feedback

breskeby committed:
Expose useDra flag as extension property

elasticsearchmachine commented:
Pinging @elastic/es-delivery (Team:Delivery)

breskeby committed:
Add generating tableau connector
